src/commands/cmd_geo.cc

```diff
@@ -355,6 +364,249 @@ class CommandGeoRadius : public CommandGeoBase {
+        with_coord_ = true;
+      } else if (parser.EatEqICase("withdist")) {
+        with_dist_ = true;
+      } else if (parser.EatEqICase("withhash")) {
+        with_hash_ = true;
+      } else {
+        return {Status::RedisParseErr, "Invalid argument given"};
+      }
+    }
+
+    if (member_ != "" && longitude_ != 0 && latitude_ != 0) {
+      return {Status::RedisParseErr, "please use only one of FROMMEMBER or FROMLONLAT"};
+    }
+
+    auto s = createGeoShape();
+    if (!s.IsOK()) {
+      return s;
+    }
+    return Commander::Parse(args);
+  }
+
+  Status Execute(Server *svr, Connection *conn, std::string *output) override {
+    std::vector<GeoPoint> geo_points;
+    redis::Geo geo_db(svr->storage, conn->GetNamespace());
+
+    auto s = geo_db.Search(args_[1], geo_shape_, origin_point_type_, member_, count_, sort_, false, GetUnitConversion(),
+                           &geo_points);
+
+    if (!s.ok()) {
+      return {Status::RedisExecErr, s.ToString()};
+    }
+    *output = generateOutput(geo_points);
+
+    return Status::OK();
+  }
+
+ protected:
+  double radius_ = 0;
+  double height_ = 0;
+  double width_ = 0;
+  int count_ = 0;
+  double longitude_ = 0;
+  double latitude_ = 0;
+  std::string member_;
+  std::string key_;
+  DistanceSort sort_ = kSortNone;
+  GeoShapeType shape_type_ = NONE;
+  OriginPointType origin_point_type_ = kNone;
+  GeoShape geo_shape_;
+
+  Status setShapeType(GeoShapeType shape_type) {
+    if (shape_type_ != NONE) {
+      return {Status::RedisParseErr, "please use only one of BYBOX or BYRADIUS"};
+    }
+    shape_type_ = shape_type;
+    return Status::OK();
+  }
+
+  Status setOriginType(OriginPointType origin_point_type) {
+    if (origin_point_type_ != kNone) {
+      return {Status::RedisParseErr, "please use only one of FROMMEMBER or FROMLONLAT"};
+    }
+    origin_point_type_ = origin_point_type;
+    return Status::OK();
+  }
+
+  Status createGeoShape() {
+    if (shape_type_ == NONE) {
+      return {Status::RedisParseErr, "please use BYBOX or BYRADIUS"};
+    }
+    geo_shape_.type = shape_type_;
+    geo_shape_.conversion = GetUnitConversion();
+
+    if (shape_type_ == CIRCULAR) {
+      geo_shape_.radius = radius_;
+    } else {
+      geo_shape_.width = width_;
+      geo_shape_.height = height_;
+    }
+
+    if (origin_point_type_ == kLongLat) {
+      geo_shape_.xy[0] = longitude_;
+      geo_shape_.xy[1] = latitude_;
+    }
+    return Status::OK();
+  }
+
+  std::string generateOutput(const std::vector<GeoPoint> &geo_points) {
+    int result_length = static_cast<int>(geo_points.size());
+    int returned_items_count = (count_ == 0 || result_length < count_) ? result_length : count_;
+    std::vector<std::string> output;
+    output.reserve(returned_items_count);
+    for (int i = 0; i < returned_items_count; i++) {
+      auto geo_point = geo_points[i];
+      if (!with_coord_ && !with_hash_ && !with_dist_) {
+        output[i] = redis::BulkString(geo_point.member);
```

Comment:
I'm afraid this line will crash because the `reserve` function doesn't change the vector's size, it only reserves the so-called capacity.

Reply:
I think you can leave `push_back/emplace_back` here.

Reply:
ah my bad, fixing this.